scripts/code_coverage_report/generate_code_coverage_report/Sources/IncrementalCoverageReportGenerator/main.swift

```diff
@@ -0,0 +1,199 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import ArgumentParser
+import Foundation
+import Utils
+
+private enum Constants {}
+
+extension Constants {
+  static let XCRESULT_EXTENSION = "xcresult"
+  // Command to get line execution counts of a file in a xcresult bundle.
+  static let XCOV_COMMAND = "xcrun xccov view --archive --file "
+  // The pattern is to match text "line_index: execution_counts" from the
+  // outcome of xcresult bundle, e.g. "305 : 0".
+  static let LINE_EXECUTION_COUNT_PATTERN = "([0-9]+)\\s*:\\s*([0-9*]+)"
+  // A file includes all newly added lines without tests covered.
+  static let DEFAULT_UNCOVERED_LINE_REPORT_FILE_NAME = "uncovered_file_lines.json"
+}
+
+struct FileIncrementalChanges: Codable {
+  let file: String
+  let addedLines: [Int]
+  enum CodingKeys: String, CodingKey {
+    case file
+    case addedLines = "added_lines"
+  }
+}
+
+extension FileIncrementalChanges {
+  static func dataModel(_ filePath: String) throws -> [FileIncrementalChanges] {
+    guard let data = JSONParser.readJSON(of: [FileIncrementalChanges].self, from: filePath)
+    else { throw ValidationError("Badly data transform.") }
+    return data
+  }
+}
+
+struct LineCoverage: Codable {
+  var fileName: String
+  var coverage: [Int?]
+  var xcresultBundle: String
+}
+
+struct IncrementalCoverageReportGenerator: ParsableCommand {
+  @Option(
+    help: "Root path of archived files in a xcresult bundle."
+  )
+  var fileArchiveRootPath: String
+
+  @Option(help: "A dir of xcresult bundles.")
+  var xcresultDir: String
+
+  @Option(
+    help: """
+    A JSON file with changed files and added line numbers. E.g. JSON file:
+    '[{"file": "FirebaseDatabase/Sources/Api/FIRDataSnapshot.m", "added_lines": [105,106,107,108,109]},{"file": "FirebaseDatabase/Sources/Core/Utilities/FPath.m", "added_lines": [304,305,306,307,308,309]}]'
+    """,
+    transform: FileIncrementalChanges.dataModel
+  )
+  var changedFiles: [FileIncrementalChanges]
+
+  @Option(
+    help: "Uncovered line JSON file output path"
+  )
+  var uncovered_line_file_json: String = Constants.DEFAULT_UNCOVERED_LINE_REPORT_FILE_NAME
+
+  func readFile(_ fileURL: URL) -> String {
+    var fileContent = ""
+    do {
+      fileContent = try String(contentsOf: fileURL)
+    } catch {
+      print("Failed reading from URL: \(fileURL), Error: " + error.localizedDescription)
+    }
+    return fileContent
+  }
+
+  // This will transfer line executions report from a xcresult bundle to an array of LineCoverage.
+  // The output will have the file name, line execution counts and the xcresult bundle name.
+  func createLineCoverageRecord(from coverageFile: String, changedFile: String,
+                                lineCoverage: [LineCoverage]? = nil) -> [LineCoverage] {
+    // The indices of the array represent the (line index - 1), while the value is the execution counts.
+    var lineExecutionCounts: [Int?] = []
```

Comment:
Why is this an optional integer?

Reply:
We got unexecutable lines, e.g. comments, which will be `nil` here.